---
title: "Session 2: Annotation of regions"
author: "Stephany Orjuela"
date: "November 13, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Genomic Ranges

## What is and how to create a `GRanges` object

Each genomic range is described by a *chromosome name*, a *start*, an *end*, and a *strand*. 
* *start* and *end* are both 1-based positions relative to the 5â€™ end of the plus strand of the chromosome, even when the range is on the minus strand.
* *start* and *end* are both considered to be included in the interval (except when the
range is empty). 
* The *width* of the range is the number of genomic positions included in it. So *width = end - start + 1*.
* *end* is always >= *start*, except for empty ranges (a.k.a.  zero-width ranges) where *end = start - 1*.

A `GRanges` is an easy way to store this type of information. It can include three parts: `GRanges` (a container for positions or regions), metadata (an optional table of information), and `seqinfo` (information about the sequences). 

```{r}
#To install:
#source("https://bioconductor.org/biocLite.R")
#biocLite("GenomicRanges")

suppressPackageStartupMessages(library(GenomicRanges))

gr <- GRanges(seqnames = c("chr1", "chr1", "chr1"),
              ranges = IRanges(start = c(1, 8, 11),
                               end = c(5, 15, 15)),
              strand = c("+", "+", "+"))
gr

# add one meta column
gr <- GRanges(seqnames = c("chr1", "chr1", "chr1"),
              ranges = IRanges(start = c(1, 8, 11),
                               end = c(5, 15, 15)),
              strand = c("+", "+", "+"),
              type = c("exon", "intron", "exon"))
gr
```

## How to subset, split and combine `GRanges` objects

```{r}
# extract the first row
gr[1]

# extract rows that have exon type
gr[gr$type == "exon"]

# split the GRanges into a GRangesList based on the type (exon or intron)
grs <- split(gr, gr$type)
grs

# combine multiple GRanges objects into one 
gr <- c(grs[[1]], grs[[2]])
gr
```


## How to extract components from the GRanges object.

```{r}
# extract Granges
granges(gr)

# extract columns
seqnames(gr)
ranges(gr)
strand(gr)
start(gr)
end(gr)
width(gr)

# extract the metadata
mcols(gr)

# extract the seqinfo
seqinfo(gr)

```

## How to overlap two different `GRanges` objects

```{r fig.width=10}

gr2 <- GRanges(seqnames = c("chr1", "chr1"),
              ranges = IRanges(start = c(5, 16),
                               end = c(5, 20)),
              strand = c("+", "+"))

# findOverlaps
over <- findOverlaps(gr, gr2)
over

queryHits(over)
```

# Annotating regions (where is my region in the genome?)

There are many packages available to annotate regions, here we will use a new package called `annotatr`

```{r}
#To install:
#source("https://bioconductor.org/biocLite.R")
#biocLite("annotatr")

library(annotatr)
```

The package uses built-in annotations including CpG annotations, genic annotations. We will use a region defined in a `GRanges` object to determine what gene it corresponds to, and where it is within this gene.

```{r}
#My region
DMR <- GRanges("chr3", IRanges(36993332, 37050918), strand = "+")
  
#CpG annotations
annotscpg <- c("hg19_cpg_islands", "hg19_cpg_shores", "hg19_cpg_shelves", "hg19_cpg_inter")

annotations <- build_annotations(genome = 'hg19', annotations = annotscpg)

dm_annotated_cpg = annotate_regions(
  regions = DMR,
  annotations = annotations,
  ignore.strand = FALSE,
  quiet = FALSE)

dm_annotated_cpg

```

Here we use the `hg19` (or GRCh37) genome build for Human, but just remember there is a more recent genome build (`hg38`). You should always be aware of the genome build you use across the entire analysis, and mix them up.

```{r}
#genic annotations
annotsgene <- c("hg19_genes_promoters", "hg19_genes_3UTRs", "hg19_genes_introns", 
                "hg19_genes_exons", "hg19_genes_5UTRs", "hg19_genes_cds", "hg19_genes_intergenic",
                "hg19_genes_1to5kb")

annotations <- build_annotations(genome = 'hg19', annotations = annotsgene)

# dm_annotated_gene <- annotate_regions(
#   regions = DMR,
#   annotations = annotations,
#   ignore.strand = FALSE,
#   quiet = FALSE)
# 
# df_annotated_gene
```

something about the region.
You can look at other examples of annotation in the packages' [vignette](http://bioconductor.org/packages/release/bioc/vignettes/annotatr/inst/doc/annotatr-vignette.html)
