---
title: "Session 2: Annotation of regions"
author: "Stephany Orjuela"
date: "November 13, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Genomic Ranges

## What is and how to create a `GRanges` object

Each genomic range is described by a *chromosome name*, a *start*, an *end*, and a *strand*. 

 * *start* and *end* are both 1-based positions relative to the 5â€™ end of the plus strand of the chromosome, even when the range is on the minus strand.
 * *start* and *end* are both considered to be included in the interval (except when the
range is empty). 
 * The *width* of the range is the number of genomic positions included in it. So *width = end - start + 1*.
 * *end* is always >= *start*, except for empty ranges (a.k.a.  zero-width ranges) where *end = start - 1*.

A `GRanges` is an easy way to store this type of information. It can include three parts: `GRanges` (a container for positions or regions), metadata (an optional table of information), and `seqinfo` (information about the sequences). 

```{r}
#To install:
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("GenomicRanges", version = "3.8")


suppressPackageStartupMessages(library(GenomicRanges))

gr <- GRanges(seqnames = c("chr1", "chr1", "chr1"),
              ranges = IRanges(start = c(1, 8, 11),
                               end = c(5, 15, 15)),
              strand = c("+", "+", "+"))
gr

# add one meta column
gr <- GRanges(seqnames = c("chr1", "chr1", "chr1"),
              ranges = IRanges(start = c(1, 8, 11),
                               end = c(5, 15, 15)),
              strand = c("+", "+", "+"),
              type = c("exon", "intron", "exon"))
gr
```

## How to subset, split and combine `GRanges` objects

```{r}
# extract the first row
gr[1]

# extract rows that have exon type
gr[gr$type == "exon"]

# split the GRanges into a GRangesList based on the type (exon or intron)
grs <- split(gr, gr$type)
grs

# combine multiple GRanges objects into one 
gr <- c(grs[[1]], grs[[2]])
gr
```


## How to extract components from the GRanges object.

```{r}
# extract Granges
granges(gr)

# extract columns
seqnames(gr)
ranges(gr)
strand(gr)
start(gr)
end(gr)
width(gr)

# extract the metadata
mcols(gr)

# extract the seqinfo
seqinfo(gr)

```


# Annotating regions (where is my region in the genome?)

In some exploratory analysis, we tipically obtain a genomic region without any prior knowledge of it. In other cases, if we know that the region is to close a gene of interest, we might be want to know if it's close to other genes, CpG islands, promoters, etc., to have a better understanding of the region's genomic environment. This is why we annotate.

There are many packages available to annotate regions, here we will use a new package called `annotatr`. [Manual here](http://bioconductor.org/packages/release/bioc/vignettes/annotatr/inst/doc/annotatr-vignette.html)


```{r}
#To install:
#BiocManager::install("annotatr", version = "3.8")

library(annotatr)
```

The package uses built-in annotations including CpG annotations and genic annotations. We will use a region defined in a `GRanges` object as input.

Here we use the `hg19` (or GRCh37) genome build for Human, but just remember there is a more recent genome build (`hg38`). You should always be aware of the genome build you use across the entire analysis, and not mix them up.

```{r}
#My region
DMR <- GRanges("chr11", IRanges(7273490, 7273590), strand = "*")

#CpG annotations
annotscpg <- c("hg19_cpg_islands", "hg19_cpg_shores", "hg19_cpg_shelves", "hg19_cpg_inter")

annotations <- build_annotations(genome = 'hg19', annotations = annotscpg)

dm_annotated_cpg = annotate_regions(
  regions = DMR,
  annotations = annotations,
  ignore.strand = TRUE,
  quiet = FALSE)

dm_annotated_cpg
```

The output is also a `GRanges` object, with a bit more complexity. To actually see the annotations we have to access the column `annot`

```{r}
dm_annotated_cpg$annot
```

Here we see this region overlaps a CpG island. Now we look at genic annotations:

```{r}
#genic annotations
annotsgene <- c("hg19_genes_promoters", "hg19_genes_3UTRs", "hg19_genes_introns", 
                "hg19_genes_exons", "hg19_genes_5UTRs", "hg19_genes_cds", "hg19_genes_intergenic",
                "hg19_genes_1to5kb")

annotations <- build_annotations(genome = 'hg19', annotations = annotsgene)

dm_annotated_gene <- annotate_regions(
  regions = DMR,
  annotations = annotations,
  ignore.strand = TRUE,
  quiet = FALSE)

dm_annotated_gene$annot
```

Here we see that the region is most likely within the SYT9 gene, however the location within the gene is a bit tricky. We see some hits for `exon` and `intron`, and 1 hit for `CDS`. This suggests the region is somewhere within the coding region of SYT9, but it's not very clear where. To determine this we can make a plot to help us interprete the regions location. For this we use the library `Gviz` which helps us plot the actual gene in an easier way.

```{r}
#BiocManager::install("Gviz", version = "3.8")
#BiocManager::install("EnsDb.Hsapiens.v75", version = "3.8")

suppressPackageStartupMessages(library(Gviz))
suppressPackageStartupMessages(library(EnsDb.Hsapiens.v75))

#Get gene locations
edb <- EnsDb.Hsapiens.v75
gr <- getGeneRegionTrackForGviz(edb, chromosome = "chr11",
                                start = 7260009, end = 7490273) #locations of the gene we found

#Turn into a Gviz track
geneTrack <- GeneRegionTrack(gr, collapseTranscripts = F, fill = "blue")

#Turn annotations into a Gviz track
aTrack <- AnnotationTrack(dm_annotated_gene$annot, group = dm_annotated_gene$annot$id)

#Turn our region into a Gviz track
myTrack <- AnnotationTrack(DMR, fill = "red")

#plot all track together
plotTracks(list(geneTrack, myTrack, aTrack), 
           from = min(start(dm_annotated_gene$annot), start(DMR)), 
           to = max(end(dm_annotated_gene$annot), end(DMR)),
           groupAnnotation = "group")
```

The biggest annotations (`intron`) make it a bit difficult to visualize. Let's take them out

```{r}

geneTrack <- GeneRegionTrack(gr, collapseTranscripts = F, fill = "blue")
aTrack <- AnnotationTrack(dm_annotated_gene$annot, group = dm_annotated_gene$annot$id)
myTrack <- AnnotationTrack(DMR, fill = "red")

plotTracks(list(geneTrack, myTrack, aTrack), 
           from = min(start(dm_annotated_gene$annot[-c(4,5,6)]), start(DMR)), 
           to = max(end(dm_annotated_gene$annot[-c(4,5,6)]), end(DMR)),
           groupAnnotation = "group")
```

We can see now that our region actually overlaps both an intron and an exon in 2 of the transcripts of gene SYT9. In the other transcripts it only overlaps the intron. This is just to ilustrate that annotating can be an arbitrary task, and in this case some people migh say the region is on the exon, others on the intron, and others might not really care as long as they know it is on gene SYT9. In the end, it depends on your question of interest. 
